Bricks/sync/owned_borrowed.h

```diff
@@ -0,0 +1,492 @@
+  T* operator->() { return &p_actual_instance_->instance_; }
+  const T* operator->() const { return &p_actual_instance_->instance_; }
+
+  // Exclusive accessors.
+  // THREAD-SAFE WITH RESPECT TO MAKING SURE UNDERLYING OBJECT IS ACCESSED EXCLUSIVELY BY THE PASSED IN FUNCTOR.
+  // NEVER THROWS.
+  void ExclusiveUse(std::function<void(T&)> f) {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    f(operator*());
+  }
+  void ExclusiveUse(std::function<void(const T&)> f) const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    f(operator*());
+  }
+  template <typename V>
+  V ExclusiveUse(std::function<V(T&)> f) {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return f(operator*());
+  }
+  template <typename V>
+  V ExclusiveUse(std::function<V(const T&)> f) const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return f(operator*());
+  }
+
+  // Return the number of registered borrower users besides the master one.
+  // For unit-testing purposes mostly, but may end up useful. -- D.K.
+  // THREAD-SAFE. NEVER THROWS.
+  size_t NumberOfActiveBorrowers() const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return p_actual_instance_->borrowers_.size();
+  }
+
+  // Return the total number of registered borrower users registered, with some possibly already out of scope.
+  // For unit-testing purposes mostly, but may end up useful. -- D.K.
+  // THREAD-SAFE. NEVER THROWS.
+  size_t TotalBorrowersSpawnedThroughoutLifetime() const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return p_actual_instance_->total_borrowers_spawned_throughout_lifetime_;
+  }
+
+ protected:
+  instance_t* p_actual_instance_;  // Is set to `nullptr` externally when moved.
+
+  void InternalUnRegister() {
+    if (key_) {
+      CURRENT_ASSERT(p_actual_instance_);
+      // `*this` is a valid (non - std::move()-d - away) `BorrowedWithCallback`.
+      // As it is being terminated, it should unregister its callback, so that:
+      // 1) If the object is still alive, to have this callback not be called upon it being destructed, and/or
+      // 2) If the object is now destructing, to indicate this borrower has released it, to decrement
+      //    the active borrowers counter, and trigger a condition variable `*p_actual_instance_` is waiting on
+      //    if this counter has reached zero with this unregistration.
+      p_actual_instance_->UnRegisterBorrower(key_);
+      key_ = 0u;
+    }
+    p_actual_instance_ = nullptr;
+  }
+
+ private:
+  size_t key_;  // The index of the slave user to mark as left the scope, or `0u` if N/A (master or moved away).
+};
+
+namespace impl {
+
+// `UniqueInstanceContainer<T>` must be a dedicated struct for `Owned<T>` to construct
+// the actual object before initializing the `WeakBorrowed<T>` associated with it.
+// `UniqueInstanceContainer<T>` is constructible, movable, and nothing else.
```

Comment:
1) I'd be reluctant to inherit from `std::unique_prtr`.
2) In `Stream` and `Storage`, for the static `Create*(*)` "constructors", `friend class UniqueInstance` is the way to go.